### PR TITLE
[java] Fix #4912 - grammar for TWR allows this expression

### LINKED ISSUE #4912: [java] Unable to parse some Java9+ resource references
**Affects PMD Version:**
7.0.0

**Description:**

Since Java 9, PMD has allowed for concise resources (ie: not variables declared within the try-with-resources), but some constructs are currently rejected by the parser even if valid Java code.

**Code Sample demonstrating the issue:**

```java
class Foo implements AutoCloseable {
      void testThis() throws Exception {
          try (this) { // unsupported
          }
      }

     void testQualifiedThis() throws Exception {
          try (Foo.this) { // unsupported
          }
      }

     void testArrayAccess(AutoCloseable[] c) throws Exception {
          try (c[0]) { // unsupported
          }
      }

      @Override
      public void close() throws Exception {
      }
}
```

Probably others too…

**Running PMD through:** Any

Probably, the saner fix would be to simply remove this check altogether, and continue under the premise that the code compiles…

https://github.com/pmd/pmd/blob/a28943ae1b2e25d3aecb08108fdf1376e23d50f7/pmd-java/etc/grammar/Java.jjt#L2868-L2873

### LINKED ISSUE #4930: [java] EmptyControlStatement should not allow empty try with concise resources
**Affects PMD Version:** 7.0.0

**Rule:** [EmptyControlStatement](https://docs.pmd-code.org/latest/pmd_rules_java_codestyle.html#emptycontrolstatement)

**Description:**
Since #432 was closed EmptyControlStatement allows empty try-with-resources, provided the resources are named with a name like `ignored`.

I think this should only be done for resources that don't use shorthand form. For instance in our tests:
https://github.com/pmd/pmd/blob/f87f24ef9075ef0139a78d3a1ce90f9a976ee601/pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/EmptyControlStatement.xml#L138-L141

This looks like an anti pattern to me. If you're going to use an empty try statement for this you should write `in.close()` because it has the same effect but is clearer.

On the other hand the following should probably still be allowed:
https://github.com/pmd/pmd/blob/f87f24ef9075ef0139a78d3a1ce90f9a976ee601/pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/EmptyControlStatement.xml#L121-L122

Here an explicit `ignored` name is given to an expression. Even if it could be written
```java
ClientResponse response = execute(() -> target.request(mediaTypes).delete(), DELETE, new ExpectedResponse(status, required)));
response.close();
```
this code would pollute the enclosing scope with a useless variable name.

Another case to keep in mind is the following:
```java
try (in) {
} catch (Foo e) {
  // do something
}
```
If there is at least one catch block then it should be ok to keep the try block empty, even if the resource is unnamed.

Probably the rule should actually not care about whether the variables are named `ignored` or the like. If the variable is not used it will be flagged by UnusedVariable already.

**Expected outcome:**

PMD should report a violation at line ..., but doesn't. This is a false-negative.

**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]*

### PR DESCRIPTION
**Pull request by oowekyala**
## Related issues

- Fix #4912
- Fixes #4930 

## Ready?

- [ ] Added unit tests for fixed bug/feature
- [ ] Passing all unit tests
- [ ] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [ ] Added (in-code) documentation (if needed)

### CODE REVIEW

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTResource.java [hunk 1]
```diff
@@ -48,24 +49,15 @@ public boolean isConciseResource() {
      * then returns the sequence of names that identifies the expression.
      * If this has a local variable declaration, then returns the name
      * of the variable.
+     *
+     * @deprecated Since 7.1.0. This method is not very useful because the expression
+     *          might be complex and not have a real "name". In this case we return the
+     *          expression pretty printed.
      */
+    @Deprecated
```
Comment (jsotuyod) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTResource.java:57]: `JavaAttributePrinter` will not print any deprecated attributes, so all tree dump tests using resources will fail with this change.

Comment (oowekyala) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTResource.java:57]: Getters that return a node are not in the tree dump tests though iirc

Comment (jsotuyod) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTResource.java:57]: But this is not returning a node. It's a string. That's why the build is failing on this pr.

Comment (adangel) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTResource.java:57]: The method `getStableName()` is also used in EmptyControlStatementRule (to exclude `unused` vars) - so, not sure whether we can simply deprecate this...

The javadoc is definitively useful to describe when to expect a "useful" value of stable name and when not...

Comment (oowekyala) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTResource.java:57]: True, although in https://github.com/pmd/pmd/issues/4930 I argue we should stop caring about the name of the variable when it is a concise resource. So we don't need a "stable name" anymore, we just need to get the name of the VariableId when the resource is like a LocalVariableDeclaration

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to master:<br />
This changeset changes 110 violations,<br />
introduces 57 new violations, 0 new errors and 0 new configuration errors,<br />
removes 5 violations, 1 errors and 0 configuration errors.<br />
<a href="https://github.com/pmd/pmd/actions/runs/8895662123?pr=4934">Download full report as build artifact</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: docs/pages/release_notes.md [hunk 2]
```diff
@@ -16,13 +16,19 @@ This is a {{ site.pmd.release_type }} release.
...
 * java-bestpractices
   * [#4278](https://github.com/pmd/pmd/issues/4278): \[java] UnusedPrivateMethod FP with Junit 5 @MethodSource and default factory method name
   * [#4852](https://github.com/pmd/pmd/issues/4852): \[java] ReplaceVectorWithList false-positive (neither Vector nor List usage) 
   * [#4975](https://github.com/pmd/pmd/issues/4975): \[java] UnusedPrivateMethod false positive when using @MethodSource on a @Nested test
 
 ### 🚨 API Changes
 
+#### Deprecated API
+
+* {% jdoc java::lang.java.ast.ASTResource#getStableName() %} and the corresponding attribute `@StableName`
```
Comment (adangel) [inline docs/pages/release_notes.md:30]:
```suggestion
* pmd-java
  * {% jdoc !!java::lang.java.ast.ASTResource#getStableName() %} and the corresponding attribute `@StableName`
```

File: docs/pages/release_notes.md [hunk 3]
```diff
@@ -16,13 +16,19 @@ This is a {{ site.pmd.release_type }} release.
 
 ### 🐛 Fixed Issues
 
+* java
+  * [#4912](https://github.com/pmd/pmd/issues/4912): \[java] Unable to parse some Java9+ resource references
 * java-bestpractices
   * [#4278](https://github.com/pmd/pmd/issues/4278): \[java] UnusedPrivateMethod FP with Junit 5 @MethodSource and default factory method name
   * [#4852](https://github.com/pmd/pmd/issues/4852): \[java] ReplaceVectorWithList false-positive (neither Vector nor List usage) 
   * [#4975](https://github.com/pmd/pmd/issues/4975): \[java] UnusedPrivateMethod false positive when using @MethodSource on a @Nested test
```
Comment (adangel) [inline docs/pages/release_notes.md:25]:
```suggestion
* java-codestyle
  * [#4930](https://github.com/pmd/pmd/issues/4930): \[java] EmptyControlStatement should not allow empty try with concise resources

```

**Review by adangel — APPROVED** [review]

**Pull request merged**